src/query_plan/mod.rs

```diff
@@ -1 +1,210 @@
+use apollo_compiler::executable::{Field, InlineFragment, OperationType, Selection, SelectionSet};
+use apollo_compiler::{ExecutableDocument, NodeStr};
+use std::sync::Arc;
+
 pub mod operation;
+
+pub struct QueryPlan {
+    node: Option<TopLevelPlanNode>,
+}
+
+pub enum TopLevelPlanNode {
+    Subscription(SubscriptionNode),
+    Fetch(FetchNode),
+    Sequence(SequenceNode),
+    Parallel(ParallelNode),
+    Flatten(FlattenNode),
+    Defer(DeferNode),
+    Condition(ConditionNode),
+}
+
+pub struct SubscriptionNode {
+    primary: FetchNode,
+    rest: Option<PlanNode>,
+}
+
+pub enum PlanNode {
+    Fetch(Arc<FetchNode>),
+    Sequence(Arc<SequenceNode>),
+    Parallel(Arc<ParallelNode>),
+    Flatten(Arc<FlattenNode>),
+    Defer(Arc<DeferNode>),
+    Condition(Arc<ConditionNode>),
+}
+
+pub struct FetchNode {
+    service_name: NodeStr,
+    /// Optional identifier for the fetch for defer support. All fetches of a given plan will be
+    /// guaranteed to have a unique `id`.
+    id: Option<NodeStr>,
+    /// If query planner defer support is enabled _and_ the subgrpah named `service_name` supports
+    /// defer, then this boolean says whether `operation` contains some @defer. Unset otherwise.
+    has_defers: Option<bool>,
+    variable_usages: Vec<NodeStr>,
+    /// `Selection`s in apollo-rs _can_ have a `FragmentSpread`, but this `Selection` is
+    /// specifically typing the `requires` key in a built query plan, where there can't be
+    /// `FragmentSpread`.
+    // PORT_NOTE: This was its own type in the JS codebase, but it's likely simpler to just have the
+    // constraint be implicit for router instead of creating a new type.
+    requires: Vec<Selection>,
+    // PORT_NOTE: We don't serialize the "operation" string in this struct, as these query plan
+    // nodes are meant for direct consumption by router (without any serdes), so we leave the
+    // question of whether it needs to be serialized to router.
+    operation_document: ExecutableDocument,
```

Comment:
> After a document is put inside a fetch node, is it ever mutated again?

I'm not sure about whether router typically mutates this document (or whether it will in the near future). @Geal may know more here.

> Is it expected to always be valid? If no and yes respectively, this should be Valid<ExecutableDocument>.

It should be valid (though I don't think the JS codebase ever fully validates it; at best, since the JS codebase uses a representation with some built-in validations, we get those, but not the rest). Even if we're not sure about mutation, I'd say it's fine to use `Valid<ExecutableDocument>` here with `assume_valid` with a way to optionally do full validation for debugging.

> If we consider it a bug if the query planner emits invalid document, the code creating fetch nodes could validate only when debug assertions are enabled, like in https://github.com/apollographql/apollo-rs/blob/d97abbc94299c998e7e8eac762adc315917da127/crates/apollo-compiler/src/execution/introspection_split.rs#L240-L246

I was gonna suggest passing down a variable like we do for `extract_subgraphs_from_supergraph()`, but using `debug_assertions` looks better. Thanks for the tip!